### ISSUE TITLE
Dotenv values containing spaces must be surrounded by quotes, is this a problem with artisan:config:cache?

### ISSUE THREAD

**Issue by gkarugi** [lorisleiva/laravel-deployer#82]
Greetings, important of all thanks for this great package ❤️

My deployment is failing due to the error "Dotenv values containing spaces must be surrounded by quotes". All my env values don't have spaces, I have checked this severally, so I am sure of this.

Mind you, because I suspected this line ` 'artisan:config:cache'` in the config was the problem, when I comment it out everything deploys well and the deployed app is running

I am generating my env using gitlab ci. below is the gitlab ci, log and config

```
image: lorisleiva/laravel-docker:latest

.init_ssh: &init_ssh |
  eval $(ssh-agent -s)
  echo "$SSH_PRIVATE_KEY" | tr -d '\r' | ssh-add - > /dev/null
  mkdir -p ~/.ssh
  chmod 700 ~/.ssh
  [[ -f /.dockerenv ]] && echo -e "Host *\n\tStrictHostKeyChecking no\n\n" > ~/.ssh/config

  # Replace the last line with the following lines if you'd rather
# leave StrictHostKeyChecking enabled (replace yourdomain.com):
#
#  ssh-keyscan yourdomain.com >> ~/.ssh/known_hosts
#  chmod 644 ~/.ssh/known_hosts
  
.change_file_permissions: &change_file_permissions |
  find . -type f -not -path "./vendor/*" -exec chmod 664 {} \;    
  find . -type d -not -path "./vendor/*" -exec chmod 775 {} \;

.prepare_env_file: &prepare_env_file |
  echo -e "APP_ENV=production
  APP_KEY=$APP_KEY
  APP_DEBUG=false
  APP_LOG_LEVEL=debug
  APP_URL=$APP_URL
  LOG_CHANNEL=stack
  DB_CONNECTION=mysql
  DB_HOST=$DB_HOST
  DB_PORT=3306
  DB_DATABASE=$DB_DATABASE
  DB_USERNAME=$DB_USERNAME
  DB_PASSWORD=$DB_PASSWORD
  SESSION_LIFETIME=120
  QUEUE_CONNECTION=database
  MAIL_DRIVER=sparkpost
  SPARKPOST_SECRET=$SPARKPOST_SECRET
  MEDIA_DEFAULT_DISK=public
  FILESYSTEM_DRIVER=public
  AFRICAS_TALKING_USERNAME=$AFRICAS_TALKING_USERNAME
  AFRICAS_TALKING_KEY=$AFRICAS_TALKING_KEY
  ADMIN_NOTIFICATION_EMAIL=$ADMIN_NOTIFICATION_EMAIL
  ADMIN_NOTIFICATION_PHONE=$ADMIN_NOTIFICATION_PHONE
  " > .env 

composer:
  stage: build
  cache:
    key: ${CI_COMMIT_REF_SLUG}-composer
    paths:
      - vendor/
  script:
      - composer install --prefer-dist --no-ansi --no-interaction --no-progress --no-scripts
      - *prepare_env_file
      # - cp .env.example .env
      # - echo ${ENV} > .env
      - cat .env
      # - php artisan key:generate
  artifacts:
    expire_in: 1 month
    paths:
      - vendor/
      - .env
  only:
   - master

production:
  stage: deploy
  script:
    - *init_ssh
    - *change_file_permissions
    - php artisan deploy example.com -s upload
  environment:
    name: production
    url: http://example.com
  when: manual
  only:
   - master
```    
```
<?php

return [
    /*
    |--------------------------------------------------------------------------
    | Default deployment strategy
    |--------------------------------------------------------------------------
    |
    | This option defines which deployment strategy to use by default on all
    | of your hosts. Laravel Deployer provides some strategies out-of-box
    | for you to choose from explained in detail in the documentation.
    |
    | Supported: 'basic', 'firstdeploy', 'local', 'pull'.
    |
    */

    'default' => 'basic',

    /*
    |--------------------------------------------------------------------------
    | Custom deployment strategies
    |--------------------------------------------------------------------------
    |
    | Here, you can easily set up new custom strategies as a list of tasks.
    | Any key of this array are supported in the `default` option above.
    | Any key matching Laravel Deployer's strategies overrides them.
    |
    */

    'strategies' => [
    ],

    /*
    |--------------------------------------------------------------------------
    | Hooks
    |--------------------------------------------------------------------------
    |
    | Hooks let you customize your deployments conveniently by pushing tasks
    | into strategic places of your deployment flow. Each of the official
    | strategies invoke hooks in different ways to implement their logic.
    |
    */

    'hooks' => [
        // Right before we start deploying.
        'start' => [
        ],

        // Code and composer vendors are ready but nothing is built.
        'build' => [
        ],

        // Deployment is done but not live yet (before symlink)
        'ready' => [
            'artisan:storage:link',
            'artisan:view:clear',
            'artisan:cache:clear',
            'artisan:migrate',
            'artisan:config:cache',
        ],

        // Deployment is done and live
        'done' => [
            'fpm:reload',
        ],

        // Deployment succeeded.
        'success' => [
        ],

        // Deployment failed.
        'fail' => [
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Deployment options
    |--------------------------------------------------------------------------
    |
    | Options follow a simple key/value structure and are used within tasks
    | to make them more configurable and reusable. You can use options to
    | configure existing tasks or to use whithin your own custom tasks.
    |
    */

    'options' => [
        'application' => config('app.name'),
        'repository' => 'https://gkarugi:QgVC4HgxxJ1JwasK4fXx@gitlab.com/int/valuation.git',
        'php_fpm_service' => 'php7.1-fpm',
    ],

    /*
    |--------------------------------------------------------------------------
    | Hosts
    |--------------------------------------------------------------------------
    |
    | Here, you can define any domain or subdomain you want to deploy to.
    | You can provide them with roles and stages to filter them during
    | deployment. Read more about how to configure them in the docs.
    |
    */

    'hosts' => [
        'example.com' => [
            'deploy_path' => '/usr/share/nginx/html',
            'user' => 'gitlabdeployer',
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Localhost
    |--------------------------------------------------------------------------
    |
    | This localhost option give you the ability to deploy directly on your
    | local machine, without needing any SSH connection. You can use the
    | same configurations used by hosts to configure your localhost.
    |
    */

    'localhost' => [
    ],

    /*
    |--------------------------------------------------------------------------
    | Include additional Deployer recipes
    |--------------------------------------------------------------------------
    |
    | Here, you can add any third party recipes to provide additional tasks,
    | options and strategies. Therefore, it also allows you to create and
    | include your own recipes to define more complex deployment flows.
    |
    */

    'include' => [
    ],

    /*
    |--------------------------------------------------------------------------
    | Use a custom Deployer file
    |--------------------------------------------------------------------------
    |
    | If you know what you are doing and want to take complete control over
    | Deployer's file, you can provide its path here. Note that, without
    | this configuration file, the root's deployer file will be used.
    |
    */

    'custom_deployer_file' => false,
];

```                                                    
```
Error Output:
================

In Client.php line 99:
                                                                               
  The command "/usr/bin/php /usr/share/nginx/html/rele  
  ases/1/artisan config:cache" failed.                                         
                                                                               
  Exit Code: 1 (General error)                                                 
                                                                               
  Host Name: example.com                                            
                                                                               
  ================                                                             
  Configuration cache cleared!                                                 
  The environment file is invalid: Dotenv values containing spaces must be su  
  rrounded by quotes.                                                          
                                                                               

deploy [-p|--parallel] [-l|--limit LIMIT] [--no-hooks] [--log LOG] [--roles ROLES] [--hosts HOSTS] [-o|--option OPTION] [--] [<stage>]



  at /builds/int/valuation/vendor/symfony/process/Process.php:256
    252|      */
    253|     public function mustRun(callable $callback = null, array $env = array())
    254|     {
    255|         if (0 !== $this->run($callback, $env)) {
  > 256|             throw new ProcessFailedException($this);
    257|         }
    258| 
    259|         return $this;
    260|     }

  Exception trace:

  1   Symfony\Component\Process\Process::mustRun(Object(Closure))
      /builds/int/valuation/vendor/lorisleiva/laravel-deployer/src/LaravelDeployer/Commands/BaseCommand.php:110

  2   Lorisleiva\LaravelDeployer\Commands\BaseCommand::process("vendor/bin/dep --file=vendor/lorisleiva/laravel-deployer/.build/deploy.php deploy 'example.com'")
      /builds/int/valuation/vendor/lorisleiva/laravel-deployer/src/LaravelDeployer/Commands/BaseCommand.php:58

  Please use the argument -v to see more details.
ERROR: Job failed: exit code 1
```

**Comment by lorisleiva** [comment]
Hi there 👋 

Thanks I'm glad you're enjoying this Laravel package! 🙃 

Have you checked the content of your GitLab variables? Are you sure none of them contain any space? Either of these basically: 
`$APP_KEY, $APP_URL, $DB_HOST, $DB_DATABASE, $DB_USERNAME, $DB_PASSWORD, $SPARKPOST_SECRET, $AFRICAS_TALKING_USERNAME, $AFRICAS_TALKING_KEY, $ADMIN_NOTIFICATION_EMAIL, $ADMIN_NOTIFICATION_PHONE`.

To be absolutely certain the problem does not come from this `prepare_env_file` script, you can always wrap them in double quotes anyway like that `APP_URL="$APP_URL"`.

If the problem still occurs when all GitLab variables are wrapped in double quotes, then it must be that your application is not using the environment you're defining in your CI file.

**Comment by gkarugi** [comment]
Hey,

Yes, I have triple checked that there is no value with a space.... I have also wrapped all values with double quotes.

Here is the generated env file

```
$ cat > .env  <<EOF # collapsed multi-line command
APP_ENV="production"
APP_KEY="base64:Vsadsadsaqj7aXxasdasdQJi9/masdasEv55fJkl4rVrg="
APP_DEBUG=false
APP_LOG_LEVEL="debug"
APP_URL="localhost:8000"
LOG_CHANNEL="stack"
DB_CONNECTION="mysql"
DB_HOST="127.0.0.1"
DB_PORT=3306
DB_DATABASE="db"
DB_USERNAME="user"
DB_PASSWORD="password"
SESSION_LIFETIME=120
QUEUE_CONNECTION="database"
MAIL_DRIVER="sparkpost"
SPARKPOST_SECRET="dasdasdasd"
MEDIA_DEFAULT_DISK="azure"
FILESYSTEM_DRIVER="azure"
AZURE_STORAGE_NAME="storage-name"
AZURE_STORAGE_KEY="fsdfdsfdsfdsfs"
AZURE_STORAGE_CONTAINER="azure-container"
AFRICAS_TALKING_USERNAME="user"
AFRICAS_TALKING_KEY="asdasdsadasd"
ADMIN_NOTIFICATION_EMAIL="info@example.org"
ADMIN_NOTIFICATION_PHONE="254712345678"
SENTRY_LARAVEL_DSN="https://4091sdfsdd5527f8b42sdfdsfde28ba835c5645ec6f6@sentry.io/13343434372919"
```
but still giving the same error

```
➤ Executing task artisan:cache:clear
✔ Ok
➤ Executing task artisan:config:cache
➤ Executing task deploy:failed
✔ Ok
➤ Executing task deploy:unlock
✔ Ok

In Client.php line 99:
                                                                               
  The command "/usr/bin/php /usr/share/nginx/html/autoscopevaluers.co.ke/rele  
  ases/6/artisan config:cache" failed.                                         
                                                                               
  Exit Code: 1 (General error)                                                 
                                                                               
  Host Name: autoscopevaluers.co.ke                                            
                                                                               
  ================                                                             
  Configuration cache cleared!                                                 
  The environment file is invalid: Dotenv values containing spaces must be su  
  rrounded by quotes.                                                          
                                                                               

deploy [-p|--parallel] [-l|--limit LIMIT] [--no-hooks] [--log LOG] [--roles ROLES] [--hosts HOSTS] [-o|--option OPTION] [--] [<stage>]


   Symfony\Component\Process\Exception\ProcessFailedException  : The command "vendor/bin/dep --file=vendor/lorisleiva/laravel-deployer/.build/deploy.php deploy 'autoscopevaluers.co.ke'" failed.

Exit Code: 1(General error)

Working directory: /builds/autoscopeint/valuation
```

On the server where the app is deployed, the env file deployed is the same as the one above and the app is running smoothly without error, though I cannot run `config:cache`

If am certain if the .env was the problem, the deployed app should break.

**Comment by lorisleiva** [comment]
Hi, that sounds very strange. I don't know why this would be happening.

Could you create a custom task that dumps the `.env` file and add it to your hooks? It's the only way we can see what Laravel Deployer uses when executing artisan commands. Your server might be using the right one but it looks like Laravel Deployer isn't.

**Comment by gkarugi** [comment]
Hello, I am perplexed too..

Yes I can do this.  I will revert once I have the results.

**Comment by lorisleiva** [comment]
Closing this due to inactivity. If this is an issue that more users are experiencing, please feel free to open a new issue with more elaborated information. Thank you for your understanding. 🍀
